### DISCUSSION TITLE
Plugins & Ajax

### DISCUSSION THREAD

**Discussion by vipers-web-design** [less/less.js#3771] [Ideas]
Does Ajax work with plugins?
I'm trying to use fetch(), but getting the error "fetch is not defined".

Then I switch to XMLHttpRequest(), and get "XMLHttpRequest() is not defined".

What I'm trying to do is build a social network, and I'm working on the "status box placeholder", like on Facebook where you see "What's on your mind, {your firstname}?"

I have;
```
div[contenteditable]:empty::before {
  content+_: "What's on your mind,";
}
```
The ajax in a plug-in would return "{your firstname}?", and then I'd have to merge them.

**Comment by iChenLei** [comment]
https://nodejs.org/en/blog/announcements/v18-release-announce/#fetch-experimental

https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest

---------------

Both fetch() and XMLHttpRequest() are not supported out of the box by Node.js if your code runs on the Node.js runtime. You must recognize that these two functions are browser APIs. @vipers-web-design

**Comment by vipers-web-design** [comment]
This website has a few dozen functional fetch() requests already.

**Reply by vipers-web-design** [reply]
I'm a LAMP-stack dev using MAMP Pro on macOS. I don't know the tech terms, I just code.
As I said, this website already has dozens of fetch() functions, all working as they should.

This is getFirstname.js;
```js
registerPlugin({
  install: function (less, pluginManager, functions) {
    functions.add('firstname', function () {
      fetch('getFirstname.php', {
        method: 'GET'
      }).then(resp => resp.text()).then(data => {
        return data;
      });
    });
  }
});
```
I switched over to using modules, still getting the same error "fetch is not defined"
```js
module.exports = {
  install: function (less, pluginManager, functions) {
    functions.add('firstname', function () {
      fetch('getFirstname.php', {
        method: 'GET'
      }).then(resp => resp.text()).then(data => {
        return data;
      });
    });
  }
};
```

This is the Less file;
```less
@plugin 'getFirstname';
&:empty {
  &::before {
    content+_: "What's on your mind,";
    color: #888;
  }
  &::before  {
    content+_: firstname();
  }
}
```

I'm not going to show all the code in getFirstname.php, but under the Network tab in DevTools does see the script and shows me it is returning the data I need.
```php
<?php
// db conn
// select query
$row = $stmt->fetch();
echo $row[ 'fn' ] . '?';
?>
```
<img width="423" alt="Screenshot 2023-01-07 at 10 40 45 AM" src="https://user-images.githubusercontent.com/8532498/211161283-a08cbd8d-ffcb-4465-9e60-68bb48b97c4d.png">
